### ISSUE TITLE
Implementing ImageNet in R

### ISSUE THREAD

**Issue by hoaphumanoid** [dmlc/mxnet#3367]
I'm trying to implement Imagenet in R using [alexnet ](https://github.com/dmlc/mxnet/blob/master/example/image-classification/symbol_alexnet.R)architecture. I successfully made it work in python, with the exception of a couple of architectures that are [not working](https://github.com/dmlc/mxnet/issues/3359). However, alexnet is working in python.

What I get right now is that the accuracy is not going up: `0.00100694444444444, 0.00112268518518518, 0.00107638888888889, 0.00110416666666666, 0.00107638888888889...`. It goes like this forever.

I'm using 4 GPUs and a batch size of 576 (I tried different batch sizes). This is the code I have:

```R
# Train imagenet
require(mxnet)
require(argparse)

get_iterator <- function(args) {
  data.shape <- c(args$data_shape, args$data_shape, 3)
  train = mx.io.ImageRecordIter(
    path.imgrec     = file.path(args$data_dir, args$train_dataset),
    batch.size      = args$batch_size,
    data.shape      = data.shape,
    mean.r          = 123.68,
    mean.g          = 116.779,
    mean.b          = 103.939,
    rand.crop       = TRUE,
    rand.mirror     = TRUE
  )
  
  val = mx.io.ImageRecordIter(
    path.imgrec     = file.path(args$data_dir, args$val_dataset),
    batch.size      = args$batch_size,
    data.shape      = data.shape,
    mean.r          = 123.68,
    mean.g          = 116.779,
    mean.b          = 103.939,
    rand.crop       = FALSE,
    rand.mirror     = FALSE
  )
  ret = list(train=train, value=val)
}

parse_args <- function() {
  parser <- ArgumentParser(description='train an image classifer on ImageNet')
  parser$add_argument('--network', type='character', default='alexnet',
                      choices = c('alexnet', 'vgg','resnet'),
                      help = 'the cnn to use')
  parser$add_argument('--data-dir', type='character', default='../../data',
                      help='the input data directory')
  parser$add_argument('--train-dataset', type='character', default="train.rec",
                      help='train dataset name')
  parser$add_argument('--val-dataset', type='character', default="val.rec",
                      help="validation dataset name")
  parser$add_argument('--gpus', type='character',
                      help='the gpus will be used, e.g "0,1,2,3"')
  parser$add_argument('--batch-size', type='integer', default=128,
                      help='the batch size')
  parser$add_argument('--lr', type='double', default=.01,
                      help='the initial learning rate')
  parser$add_argument('--clip-gradient', type='double', default=5,
                      help='clip min/max gradient to prevent extreme value')
  parser$add_argument('--model-prefix', type='character',
                      help='the prefix of the model to load/save')
  parser$add_argument('--num-round', type='integer', default=10,
                      help='the number of iterations over training data to train the model')
  parser$add_argument('--kv-store', type='character', default='local',
                      help='the kvstore type')
  parser$add_argument('--num-examples', type='integer', default=1281167,
                      help='the number of training examples')
  parser$add_argument('--num-classes', type='integer', default=1000,
                      help='the number of classes')
  parser$add_argument('--data-shape', type='integer', default=224,
                      help='set images shape')
  parser$parse_args()
}
args <- parse_args()

# network
if (args$network == 'alexnet') {
  source("symbol_alexnet.R")
} else if (args$network == 'vgg'){
  source("symbol_vgg.R")
} else {
  source("symbol_resnet.R")
}
network <- get_symbol(args$num_classes)

# data
data <- get_iterator(args)
train <- data$train
val <- data$value

# devices
if (is.null(args$gpus)) {
  devs <- mx.cpu()  
} else {
  devs <- lapply(unlist(strsplit(args$gpus, ",")), function(i) {
    mx.gpu(as.integer(i))
  })
}

# save model
if (is.null(args$model_prefix)) {
  checkpoint <- NULL
} else {
  checkpoint <- mx.callback.save.checkpoint(args$model_prefix)
}

# learning rate scheduler
if (args$lr_factor < 1){
  epoch_size <- args$num_examples/args$batch_size
  #lr_scheduler <- mx.lr_scheduler.FactorScheduler(
  FactorScheduler <- mxnet:::FactorScheduler
  lr_scheduler <- FactorScheduler(
    step = as.integer(max(epoch_size * args$lr_factor_epoch, 1)),
    factor_val = args$lr_factor)
} else{
  lr_scheduler = NULL
}

# train
model = mx.model.FeedForward.create(
  X                  = train,
  eval.data          = val,
  ctx                = devs,
  symbol             = network,
  eval.metric        = mx.metric.accuracy,
  num.round          = args$num_round,
  learning.rate      = args$lr,
  momentum           = 0.9,
  wd                 = 0.00001,
  kvstore            = args$kv_store,
  array.batch.size   = args$batch_size,
  clip_gradient      = args$clip_gradient,
  lr_scheduler       = lr_scheduler,
  epoch.end.callback = checkpoint,
  initializer        = mx.init.Xavier(factor_type="in", magnitude=2),
  batch.end.callback = mx.callback.log.train.metric(50))
```
Any ideas?

**Comment by mli** [comment]
are you double checked the data? namely you can train alexnet on the .rec data using python.

**Comment by hoaphumanoid** [comment]
Yes, the same data. Samples are going at 234.11 samples/sec and accuracies are going up steadily. There are several differences between the [python ](https://github.com/dmlc/mxnet/blob/master/example/image-classification/train_imagenet.py)and R script, such as, in the iterator:
```python
        num_parts   = kv.num_workers,
        part_index  = kv.rank
```
Also in R I don't have the `lr-factor` and `clip-gradient` as in [python](https://github.com/dmlc/mxnet/blob/master/example/image-classification/train_model.py#L59). But that shouldn't be the reason right?

**Comment by mli** [comment]
clip-gradient is important to avoid very bad things.

**Comment by hoaphumanoid** [comment]
Tried to add clip_gradient. I'm not sure how to implement it in R. I added this to the arguments:
```R
  parser$add_argument('--clip-gradient', type='double', default=5,
                      help='clip min/max gradient to prevent extreme value')
```
And also this as an argument to the training function `mx.model.FeedForward.create`:
```R
  clip_gradient      = args$clip_gradient,
```

Now, I don't get any different result, the train accuracy is still not converging and the trainingis getting slower. Any other idea? Thanks

**Comment by hoaphumanoid** [comment]
I updated also the `lr_scheduler` like this:
```R
# learning rate scheduler
if (args$lr_factor < 1){
  epoch_size <- args$num_examples/args$batch_size
  #lr_scheduler <- mx.lr_scheduler.FactorScheduler(
  FactorScheduler <- mxnet:::FactorScheduler
  lr_scheduler <- FactorScheduler(
    step = as.integer(max(epoch_size * args$lr_factor_epoch, 1)),
    factor_val = args$lr_factor)
} else{
  lr_scheduler = NULL
}
```
I had to take the function `FactorScheduler` from the namespace because it wasn't in the namespace.

However, the code is not working. I realised that if I remove the scheduler in python, the training doesn't converge. 

Any idea what could be the problem?

**Comment by hoaphumanoid** [comment]
It looks that if you remove remove the num_parts = kv.num_workers and part_index = kv.rank in the image iterator in python, the training doesn't converge. 
In python it's easy to access `num_workers` and `rank`, they have [getters](https://github.com/dmlc/mxnet/blob/5ff545f2345f9b607b81546a168665bd63d02d9f/python/mxnet/kvstore.py#L272). However, in R there is only a call to a [cpp function](https://github.com/dmlc/mxnet/blob/5ff545f2345f9b607b81546a168665bd63d02d9f/R-package/R/kvstore.R).

Do you know @mli @thirdwing how I can access `num_workers` and `rank` in R?

**Comment by hoaphumanoid** [comment]
I have imagenet running in R. There are still many difference with python mostly because the python API is more developed. As soon as I have some results I'll do a PR.

**Comment by thirdwing** [comment]
Hi, Miguel,

Can you show an example? I just finished one paper so I think I have 
some time this week.

**Comment by hoaphumanoid** [comment]
Ey, KK, I just sent you an email
